File: src/main/java/com/cuupa/mailprocessor/services/ExternSemanticService.java
package com.cuupa.mailprocessor.services;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import org.springframework.http.ResponseEntity;
import org.springframework.web.client.RestTemplate;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

public class ExternSemanticService {

    private final RestTemplate restTemplate;

    private final String semanticUrl;

    private final Gson gson = new Gson();

    public ExternSemanticService(final RestTemplate restTemplate, final String semanticUrl){
        this.restTemplate = restTemplate;
        this.semanticUrl = semanticUrl;
    }

    public List<SemanticResult> getSemanticResult(final String plainText) {
        final ResponseEntity<String>
                responseEntity =
                restTemplate.postForEntity(semanticUrl, plainText, String.class);
        if(responseEntity.getStatusCode().is2xxSuccessful()){
            Type listType = new TypeToken<ArrayList<SemanticResult>>() {
            }.getType();
            return gson.fromJson(responseEntity.getBody(), listType);
        } else {
            return new ArrayList<>();
        }
    }
}

File: src/main/java/com/cuupa/mailprocessor/services/semantic/ExternSemanticService.kt
package com.cuupa.mailprocessor.services.semantic

import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import org.springframework.web.client.ResourceAccessException
import org.springframework.web.client.RestTemplate
import java.lang.reflect.Type
import java.util.*

/**
 * Calls the extern semantic serivce to classify the text
 */
class ExternSemanticService(private val restTemplate: RestTemplate, private val semanticUrl: String) {

    private val gson = Gson()

    fun getSemanticResult(plainText: String?): List<SemanticResult> {
        return try {
            val responseEntity = restTemplate.postForEntity(semanticUrl, plainText, String::class.java)
            return when {
                responseEntity.statusCode.is2xxSuccessful -> gson.fromJson(responseEntity.body, listType)
                else -> listOf()
            }
        } catch (exception: ResourceAccessException) {
            listOf(SemanticResult("OTHER", "UNKNOWN"))
        }
    }

    companion object{
        val listType: Type = object : TypeToken<ArrayList<SemanticResult?>?>() {}.type
    }
}